runtime/permissions/mod.rs

```diff
@@ -226,22 +228,267 @@ impl AsRef<str> for EnvVarName {
+  ) -> PermissionState {
+    if self.is_denied(desc) || self.is_no_prompt(desc) {
+      PermissionState::Denied
+    } else if self.is_granted(desc) {
+      match allow_partial {
+        Some(true) => PermissionState::Granted,
+        Some(false) => {
+          if self.is_partial_denied(desc) {
+            PermissionState::Denied
+          } else {
+            PermissionState::Granted
+          }
+        }
+        None => {
+          if self.is_partial_denied(desc) {
+            PermissionState::GrantedPartial
+          } else {
+            PermissionState::Granted
+          }
+        }
+      }
+    } else if allow_partial == Some(false) && self.is_partial_denied(desc) {
+      PermissionState::Denied
+    } else {
+      PermissionState::Prompt
+    }
+  }
+
+  fn request_desc(
+    &mut self,
+    desc: &Option<T>,
+    get_display_name: impl Fn() -> Option<String>,
+  ) -> PermissionState {
+    let state = self.query_desc(desc, None);
+    if state == PermissionState::Granted {
+      self.insert_granted(desc.clone());
+      return state;
+    }
+    if state != PermissionState::Prompt {
+      return state;
+    }
+    let mut message = String::with_capacity(40);
+    message.push_str(&format!("{} access", T::type_name()));
+    match get_display_name() {
+      Some(display_name) => {
+        message.push_str(&format!(" to \"{}\"", display_name))
+      }
+      None => match desc {
+        Some(desc) => message.push_str(&format!(" to \"{}\"", desc.name())),
+        None => {}
+      },
+    }
+    match permission_prompt(
+      &message,
+      T::type_name(),
+      Some("Deno.permissions.request()"),
+      true,
+    ) {
+      PromptResponse::Allow => {
+        self.insert_granted(desc.clone());
+        PermissionState::Granted
+      }
+      PromptResponse::Deny => {
+        self.insert_no_prompt(desc.clone());
+        PermissionState::Denied
+      }
+      PromptResponse::AllowAll => {
+        self.insert_granted(None);
+        PermissionState::Granted
+      }
+    }
+  }
+
+  fn revoke_desc(&mut self, desc: &Option<T>) -> PermissionState {
+    match desc.as_ref() {
+      Some(desc) => self.granted_list.retain(|v| !v.stronger_than(desc)),
+      None => {
+        self.granted_global = false;
+        // Revoke global is a special case where the entire granted list is
+        // cleared. It's inconsistent with the granular case where only
+        // descriptors stronger than the revoked one are purged.
+        self.granted_list.clear();
+      }
+    }
+    self.query_desc(desc, None)
+  }
+
+  fn is_granted(&self, desc: &Option<T>) -> bool {
+    Self::list_contains(desc, self.granted_global, &self.granted_list)
+  }
+
+  fn is_denied(&self, desc: &Option<T>) -> bool {
+    Self::list_contains(desc, self.denied_global, &self.denied_list)
+  }
+
+  fn is_no_prompt(&self, desc: &Option<T>) -> bool {
+    Self::list_contains(desc, self.no_prompt_global, &self.no_prompt_list)
+  }
```

Comment:
My mistake in the implementation I gave, `no_prompt` should be an downwards recursive check
```suggestion
  fn is_no_prompt(&self, desc: &Option<T>) -> bool {
    match desc.as_ref() {
      Some(desc) => self.no_prompt_list.iter().any(|v| desc.stronger_than(v)),
      None => self.no_prompt_global || !self.no_prompt_list.is_empty(),
    }
  }
```